File: worker/test/smoke.py
"""Same smoke test for a Mac without node 22 (python websockets)."""
import asyncio, json, sys, urllib.request
BASE = __import__('os').environ.get('BASE', 'http://127.0.0.1:18787'); WS = BASE.replace("http", "ws")
import websockets
def post(path, body):
    req = urllib.request.Request(BASE + path, data=json.dumps(body).encode(), headers={"content-type": "application/json"}, method="POST")
    return json.load(urllib.request.urlopen(req))
def get(path): return json.load(urllib.request.urlopen(BASE + path))
def ok(c, m):
    if not c: print("FAIL:", m); sys.exit(1)
async def main():
    r1 = post("/v1/join", {"mode": "pack"})["roomId"]; r2 = post("/v1/join", {"mode": "pack"})["roomId"]
    ok(r1 == r2, "same room"); race = post("/v1/join", {"mode": "race"})["roomId"]; ok(race.startswith("race-"), "race room")
    async with websockets.connect(f"{WS}/v1/rooms/{r1}/ws?id=a1&name=Sly&city=New%20Jersey") as a, \
               websockets.connect(f"{WS}/v1/rooms/{r1}/ws?id=b2&name=Marcus&city=NYC") as b:
        await a.recv(); await b.recv()
        await a.send(json.dumps({"type": "telemetry", "elapsed": 61, "metrics": {"steps": {"v": 120, "src": "pedometer"}, "spm": {"v": 72, "src": "pedometer"}, "level": {"v": 7, "src": "manual"}, "hr": {"v": 140, "src": "watch_hr"}, "bogus": {"v": 1, "src": "x"}}}))
        await b.send(json.dumps({"type": "telemetry", "elapsed": 30, "metrics": {"steps": {"v": 40, "src": "nope"}, "spm": {"v": 68, "src": "estimated", "model": "stairmaster-generic-v1"}, "level": {"v": 6, "src": "manual"}}}))
        last = None
        for _ in range(6):
            m = json.loads(await asyncio.wait_for(b.recv(), 2))
            if m["type"] == "snapshot": last = m
            if last and len(last["climbers"]) == 2 and all(c["metrics"] for c in last["climbers"]): break
        ok(last and len(last["climbers"]) == 2, "two climbers")
        sly = next(c for c in last["climbers"] if c["id"] == "a1"); marcus = next(c for c in last["climbers"] if c["id"] == "b2")
        ok(sly["metrics"]["steps"]["src"] == "pedometer" and sly["metrics"]["hr"]["v"] == 140, "provenance kept")
        ok("bogus" not in sly["metrics"], "unknown keys dropped")
        ok(marcus["metrics"]["steps"]["src"] == "estimated" and marcus["metrics"]["spm"]["model"] == "stairmaster-generic-v1", "coerced + model kept")
        ok(last["pack"]["size"] == 2 and last["pack"]["avgSpm"] == 70, f"avg spm {last['pack']['avgSpm']}")
        await a.send(json.dumps({"type": "cheer", "text": "Keep going!"}))
        m = json.loads(await asyncio.wait_for(b.recv(), 2)); ok(m["type"] == "cheer" and m["from"] == "Sly", "cheer relayed")
        live = get("/v1/live"); ok(live["climbing"] == 2, f"live 2 got {live['climbing']}")
        ok(any(c.get("city") == "NYC" for c in live["climbers"]), "city on board")
        ok(next(r for r in live["rooms"] if r["id"] == r1)["count"] == 2, "room count in lobby")
        await a.send(json.dumps({"type": "finish"}))
        for _ in range(4):
            m = json.loads(await asyncio.wait_for(b.recv(), 2))
            if m["type"] == "snapshot" and next(c for c in m["climbers"] if c["id"] == "a1")["finished"]: break
        else: ok(False, "finisher on wall")
        live2 = get("/v1/live"); ok(live2["climbing"] == 1, f"live after finish {live2['climbing']}")
    print("smoke ok:", r1, race, "avg spm", last["pack"]["avgSpm"])
asyncio.run(main())

File: worker/test/smoke.mjs
// Smoke test against a running `wrangler dev` (node 22+: global WebSocket).
// Verifies: join creates a room, two climbers land in the same pack, telemetry
// keeps its provenance, the live board sees both, finish lands on the wall.
const BASE = process.env.BASE || "http://127.0.0.1:18787";
const WS = BASE.replace(/^http/, "ws");
const assert = (c, m) => { if (!c) { console.error("FAIL:", m); process.exit(1); } };
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

const join = async (mode) => (await (await fetch(`${BASE}/v1/join`, { method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify({ mode }) })).json()).roomId;

const r1 = await join("pack");
const r2 = await join("pack");
assert(r1 === r2, `second joiner should land in the open room (${r1} vs ${r2})`);
const race = await join("race");
assert(race !== r1 && race.startsWith("race-"), "race joins its own room");

function climber(room, id, name, city) {
  const ws = new WebSocket(`${WS}/v1/rooms/${room}/ws?id=${id}&name=${name}&city=${city}`);
  const state = { snaps: [], cheers: [] };
  ws.onmessage = (e) => { const m = JSON.parse(e.data); if (m.type === "snapshot" || m.type === "hello") state.snaps.push(m); if (m.type === "cheer") state.cheers.push(m); };
  return new Promise((res) => (ws.onopen = () => res({ ws, state })));
}

const a = await climber(r1, "a1", "Sly", "New%20Jersey");
const b = await climber(r1, "b2", "Marcus", "NYC");
await sleep(300);
a.ws.send(JSON.stringify({ type: "telemetry", elapsed: 61, metrics: { steps: { v: 120, src: "pedometer" }, spm: { v: 72, src: "pedometer" }, level: { v: 7, src: "manual" }, hr: { v: 140, src: "watch_hr" }, effort: { v: 71, src: "watch_hr" }, bogus: { v: 1, src: "x" } } }));
b.ws.send(JSON.stringify({ type: "telemetry", elapsed: 30, metrics: { steps: { v: 40, src: "nope" }, spm: { v: 68, src: "estimated", model: "stairmaster-generic-v1" }, level: { v: 6, src: "manual" } } }));
await sleep(400);

const last = b.state.snaps.at(-1);
assert(last && last.climbers.length === 2, "both climbers in the snapshot");
const sly = last.climbers.find((c) => c.id === "a1");
assert(sly.metrics.steps.src === "pedometer" && sly.metrics.hr.v === 140, "provenance preserved");
assert(!("bogus" in sly.metrics), "unknown metric keys dropped");
const marcus = last.climbers.find((c) => c.id === "b2");
assert(marcus.metrics.steps.src === "estimated", "unknown source coerced to estimated");
assert(marcus.metrics.spm.model === "stairmaster-generic-v1", "model name kept");
assert(last.pack.size === 2 && last.pack.avgSpm === 70, `pack avg spm 70, got ${last.pack.avgSpm}`);

a.ws.send(JSON.stringify({ type: "cheer", text: "Keep going!" }));
await sleep(300);
assert(b.state.cheers.length === 1 && b.state.cheers[0].from === "Sly", "cheer relayed to the other climber");
assert(a.state.cheers.length === 0, "cheer not echoed to sender");

const live = await (await fetch(`${BASE}/v1/live`)).json();
assert(live.climbing === 2, `live board counts 2, got ${live.climbing}`);
assert(live.climbers.find((c) => c.city === "NYC"), "city visible on the board");
assert(live.rooms.find((r) => r.id === r1)?.count === 2, "room count reported to lobby");

a.ws.send(JSON.stringify({ type: "finish" }));
await sleep(400);
const wall = b.state.snaps.at(-1);
assert(wall.climbers.find((c) => c.id === "a1").finished === true, "finisher stays on the wall");
const live2 = await (await fetch(`${BASE}/v1/live`)).json();
assert(live2.climbing === 1, `finished climber leaves the live count, got ${live2.climbing}`);

a.ws.close(); b.ws.close();
console.log("smoke ok: rooms", r1, race, "| pack avg spm", last.pack.avgSpm);
process.exit(0);
